### PR TITLE
[Core] Generalize Encoder-Decoder `seq_lens` computation to avoid Whisper hardcoded logic  

### PR DESCRIPTION
**Pull request by NickLucche**
### Overview
This PR generalizes current encoder-decoder `attn_metadata.seq_lens` computation in order to fix some hard-coded assumptions that are only valid for Whisper and based on static config values, rather than scheduler output.

Currently, `GPUModelRunner._get_encoder_seq_lens` returns the number of scheduled encoder tokens, which is used to compute the slot_mapping needed when caching the cross attn kv cache. 
When no encoder tokens are scheduled, subsequent decoder step are able to attend to the previously cached encoder tokens by re-creating a fixed `attn_metadata.seq_lens` that is always of the same len (`max_encoder_len`) https://github.com/vllm-project/vllm/blob/6fb0215eee44cf5e4b28f57e6739ef4a51945127/vllm/attention/layers/cross_attention.py#L99-L104 
This only works because whisper encoder input size is fixed (padded or truncated to 1500), so `seq_lens` is basically immutable.

A more correct and general formulation would be that of returning the actual `encoder_seq_lens` to attend to on each step, while skipping the cross_slot_mapping computation for every step but the first decoder one (when we cache KV).
This is the gist of this PR that executes on that idea in 2 steps:
 - modifies `GPUModelRunner._get_encoder_seq_lens` to always return the number of tokens that were input to the encoder for each running request.
 - modifies `CrossAttentionBuilder.build` to skip slot_mapping computation on steps after the first decode one. Additionally, re-uses the newly computed `encoder_seq_lens` as **actual** `attn_metadata.seq_lens`, thus solving the need to create a fix `attn_metadata.seq_lens` that depends on a static config instead of scheduled tokens. 

This change will also enable OOT implementations of encoder-decoder models.

### Test with

No difference should be detected when running Whisper, the behavior should be unchanged as well as the produced attn_metadata.

Some exhaustive whisper tests:
```
pytest -v -s -x tests/entrypoints/openai/test_transcription_validation.py
```

cc @DarkLight1337 @russellb

### CODE REVIEW

File: vllm/attention/layers/cross_attention.py [hunk 1]
```diff
@@ -93,23 +84,30 @@ def build(
...
+                num_tokens = common_attn_metadata.num_computed_tokens_cpu.numpy()
+                new_metadata.encoder_seq_lens_cpu = np.where(
+                    num_tokens > 0, 0, new_metadata.encoder_seq_lens_cpu
+                )
+
+            # seq_lens is provided by model runner: initial encoder input length is
+            # needed here to know how many tokens to attend to from the cached
+            # cross-attention KV cache.
+            new_metadata.seq_lens = common_attn_metadata.encoder_seq_lens
+            new_metadata.seq_lens_cpu = common_attn_metadata.encoder_seq_lens_cpu
```
[the remark this answers is not in the document]
Comment (russellb) [reply vllm/attention/layers/cross_attention.py]: The type mismatch seems accurate, though I haven't validated its impact analysis

[the remark this answers is not in the document]
Comment (NickLucche) [reply vllm/attention/layers/cross_attention.py]: good bot

**Review by ekagra-ranjan — APPROVED** [review]
LGTM! Thanks @NickLucche for adding the relevant part from my [PR](https://github.com/vllm-project/vllm/pull/29278) in this PR. I appreciate that!

**Review by DarkLight1337 — APPROVED** [review]

File: vllm/attention/layers/cross_attention.py [hunk 2]
```diff
@@ -93,23 +84,32 @@ def build(
...
             new_metadata.max_seq_len = max_encoder_len
-
-            new_metadata.seq_lens = torch.full(
-                (new_metadata.num_reqs,),
-                max_encoder_len,
-                dtype=torch.int32,
-                device=self.device,
+            # Any computed tokens indicated decode step>1 (no chunked prefill)
+            num_cache_decodes = (
+                (common_attn_metadata.num_computed_tokens_cpu > 0).sum().item()
```
Comment (LucasWilkinson) [inline vllm/attention/layers/cross_attention.py:91]: nit: `torch.any` might be clearer/faster

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -2726,7 +2742,7 @@ def execute_model(
                         ubatch_slices=ubatch_slices,
                         logits_indices=logits_indices,
                         use_spec_decode=use_spec_decode,
-                        scheduled_encoder_inputs=scheduler_output.scheduled_encoder_inputs,
+                        num_scheduled_tokens=scheduler_output.num_scheduled_tokens,
```
Comment (LucasWilkinson) [inline vllm/v1/worker/gpu_model_runner.py:2847]: im not super familiar with `scheduler_output.scheduled_encoder_inputs` is this safe? cc @russellb

**Review by LucasWilkinson — APPROVED** [review]
Overall LGTM; but would be good to have @russellb look too
